packages/cqrs/lib/src/cqrs_result.dart

```diff
@@ -0,0 +1,202 @@
+// Copyright 2023 LeanCode Sp. z o.o.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'package:equatable/equatable.dart';
+
+import 'command_result.dart';
+import 'cqrs_error.dart';
+
+/// Extension adding [isInvalid] getter for validation error checking on
+/// [CqrsCommandResult] with error type [CqrsError].
+extension CqrsCommandResultValidationErrorExtension
+    on CqrsCommandResult<CqrsError> {
+  /// Whether is a [CqrsCommandFailure] with validation errors.
+  bool get isInvalid => error == CqrsError.validation;
+}
+
+/// Generic result for CQRS query result. Can be either [CqrsQuerySuccess]
+/// or [CqrsQueryFailure].
+sealed class CqrsQueryResult<T, E> extends Equatable {
+  /// Creates a [CqrsQueryResult] class.
+  const CqrsQueryResult();
+
+  /// Casts to [CqrsQuerySuccess] or `null`.
+  CqrsQuerySuccess<T, E>? get asSuccess => switch (this) {
+        final CqrsQuerySuccess<T, E> s => s,
+        CqrsQueryFailure() => null,
+      };
+
+  /// Casts to [CqrsQueryFailure] or `null`.
+  CqrsQueryFailure<T, E>? get asFailure => switch (this) {
+        CqrsQuerySuccess() => null,
+        final CqrsQueryFailure<T, E> f => f,
+      };
+
+  /// Whether this instance can be casted to [CqrsQuerySuccess].
+  bool get isSuccess => asSuccess != null;
+
+  /// Whether this instance can be casted to [CqrsQueryFailure].
+  bool get isFailure => asFailure != null;
+
+  /// Returns [data] of type [T] for [CqrsQuerySuccess] and `null`
+  /// for [CqrsQueryFailure].
+  T? get data => asSuccess?.data;
+
+  /// Returns [error] of type [E] for [CqrsQueryFailure] and `null`
+  /// for [CqrsQuerySuccess].
+  E? get error => asFailure?.error;
```

Comment:
Should error be generic? Isn't it always CqrsError?

Reply:
That what I was thinking about too and I'm probably going to remove generic type for error.

Reply:
After all I removed the generic error type.